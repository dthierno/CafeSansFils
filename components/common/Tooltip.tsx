import React from "react";
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Pressable,
  Animated,
} from "react-native";
import { ChevronDown, Circle, LucideIcon } from "lucide-react-native";

// Constants
import COLORS from "@/constants/Colors";
import SPACING from "@/constants/Spacing";
import TYPOGRAPHY from "@/constants/Typography";

import { useModal } from "../layouts/GlobalModal";
import FilterModalLayout from "../layouts/FilterModalLayout";

type TooltipProps = {
  /** Text label to display inside the tooltip */
  label: string;

  /** Optional status indicator to determine the icon's color and behavior */
  status?: "green" | "orange" | "red";

  /** Optional custom icon to display at the start of the tooltip */
  Icon?: LucideIcon;

  /** Whether to show the chevron icon at the end of the tooltip (default: true) */
  showChevron?: boolean;

  /** Whether to change the tooltip color when pressed (default: false) */
  changeColorOnPress?: boolean;

  /** Optional color to apply to the tooltip */
  color?: string;

  /** Optional text color to apply to the tooltip */
  textColor?: string;

  /** Optional children to display inside the tooltip */
  children?: React.ReactNode;

  /** Whether to animate the circle icon when status is green or orange (default: false) */
  animateCircle?: boolean;

  /** Callback function triggered when the tooltip is pressed */
  onPress?: () => void;
};

/**
 * ## Tooltip
 *
 * A versatile tooltip component for displaying labels, optional status indicators,
 * and interactive chevrons. The tooltip supports custom icons, dynamic statuses,
 * and press interactions. When `showChevron` is enabled, you'll need to provide
 * a children component to display inside the modal.
 *
 * ### Features
 * - Customizable status (`green`, `orange`, `red`) with default circular icons.
 * - Optional chevron icon to indicate expandable content or actions.
 * - Supports custom icons for further flexibility.
 * - Pressable for interactions (e.g., navigating or triggering actions).
 *
 * ### Example Usage
 *
 * **Basic Example**
 * ```tsx
 * <Tooltip
 *   label="Open"
 *   status="green"
 * />
 * ```
 *
 * **With Custom Icon**
 * ```tsx
 * import { Globe } from "lucide-react-native";
 *
 * <Tooltip
 *   label="Global"
 *   Icon={Globe}
 *   onPress={() => console.log("Tooltip pressed!")}
 * />
 * ```
 *
 * **Without Chevron**
 * ```tsx
 * <Tooltip
 *   label="Details"
 *   status="orange"
 *   showChevron={false}
 * />
 * ```
 *
 * @param TooltipProps - Props for the Tooltip component.
 * @returns {JSX.Element} The rendered tooltip component.
 */
export default function Tooltip({
  children,
  label,
  status,
  Icon = status && Circle, // Default to Circle if a status is provided
  showChevron = true,
  color,
  textColor,
  changeColorOnPress = false,
  animateCircle = true,
  onPress,
}: TooltipProps): JSX.Element {
  const [isPressed, setIsPressed] = React.useState(false);
  const pulseAnim = React.useRef(new Animated.Value(1)).current;

  React.useEffect(() => {
    if (animateCircle && (status === "green" || status === "orange")) {
      Animated.loop(
        Animated.sequence([
          Animated.timing(pulseAnim, {
            toValue: 1.5,
            duration: 1200,
            useNativeDriver: true,
          }),
          Animated.timing(pulseAnim, {
            toValue: 1,
            duration: 1200,
            useNativeDriver: true,
          }),
        ])
      ).start();
    }
  }, [status, animateCircle]);

  const modalContext = useModal();
  const openModal = modalContext ? modalContext.openModal : () => {};
  const closeModal = modalContext ? modalContext.closeModal : () => {};

  function handleApplyFilter() {
    closeModal();
    setIsPressed(true);
    onPress && onPress();
  }

  function handleResetFilter() {
    closeModal();
    setIsPressed(false);
  }

  function handlePress() {
    !showChevron && setIsPressed(!isPressed);

    if (showChevron) {
      openModal(
        <View>
          <FilterModalLayout
            title={label}
            handleApplyFilter={handleApplyFilter}
            handleResetFilter={handleResetFilter}
          >
            {/* Forms elements for the filter like search box, multi-select */}
            {children}
          </FilterModalLayout>
        </View>
      );
    }
    if (onPress && !showChevron) onPress();
  }
  return (
    <TouchableOpacity
      onPress={handlePress}
      style={[
        styles.tooltipContainer,
        changeColorOnPress
          ? isPressed
            ? { backgroundColor: COLORS.black }
            : { backgroundColor: COLORS.lightGray }
          : { backgroundColor: COLORS.lightGray },
        color && { backgroundColor: color },
      ]}
      activeOpacity={0.7}
      testID="tooltip-container"
    >
      {/* Icon if provided */}
      {Icon && (
        <View style={styles.iconContainer}>
          <View style={{ position: 'relative' }}>
            {animateCircle && (status === "green" || status === "orange") && (
              <Animated.View
                style={[
                  styles.pulsingDot,
                  {
                    transform: [{ scale: pulseAnim }],
                    backgroundColor:
                      status === "green"
                        ? COLORS.status.green
                        : COLORS.status.orange,
                  },
                ]}
              />
            )}
            {status === "green" ? (
              <Icon
                width={9}
                height={9}
                strokeWidth={3}
                color={COLORS.status.green}
                fill={COLORS.status.green}
                testID="tooltip-icon"
              />
            ) : status === "orange" ? (
              <Icon
                width={9}
                height={9}
                strokeWidth={3}
                color={COLORS.status.orange}
                fill={COLORS.status.orange}
                testID="tooltip-icon"
              />
            ) : status === "red" ? (
              <Icon
                width={12}
                height={12}
                strokeWidth={3}
                color={COLORS.status.red}
                fill={COLORS.status.red}
                testID="tooltip-icon"
              />
            ) : (
              <Icon
                width={14}
                height={14}
                strokeWidth={3}
                color={
                  changeColorOnPress
                    ? isPressed
                      ? COLORS.white
                      : COLORS.black
                    : COLORS.black
                }
                testID="tooltip-icon"
              />
            )}
          </View>
        </View>
      )}

      {/* Label */}
      <Text
        style={[
          TYPOGRAPHY.body.normal.semiBold,
          changeColorOnPress
            ? isPressed
              ? { color: COLORS.white }
              : { color: COLORS.black }
            : { color: COLORS.black },
          !showChevron && { marginRight: SPACING.xxs }, // Adjust margin if no chevron
          textColor && { color: textColor },
        ]}
      >
        {label}
      </Text>

      {/* Chevron Down if enabled */}
      {showChevron && (
        <ChevronDown
          width={16}
          height={16}
          strokeWidth={2.8}
          color={isPressed ? COLORS.white : COLORS.black}
          testID="chevron-down"
        />
      )}
    </TouchableOpacity>
  );
}

function InfoModalLayout({
  title,
  description = "Lorem ipsum dolor sit amet consectetur adipisicing elit. Atque nostrum accusantium autem eos vero consequatur reiciendis quae tenetur possimus sit!",
}: {
  title?: string;
  description?: string;
}) {
  return (
    <>
      <Text style={[TYPOGRAPHY.heading.medium.bold, { textAlign: "center" }]}>
        {title}
      </Text>
      <Text
        style={[
          TYPOGRAPHY.body.large.medium,
          { textAlign: "center", margin: SPACING.md },
        ]}
      >
        {description}
      </Text>
    </>
  );
}

// Styles
const styles = StyleSheet.create({
  tooltipContainer: {
    flexDirection: "row",
    alignItems: "center",
    paddingHorizontal: SPACING.md,
    paddingVertical: 10,
    borderRadius: 500, // Rounded appearance
    gap: SPACING.xxs, // Space between elements
  },
  iconContainer: {
    marginRight: SPACING.xxs, // Space between icon and label
  },
  pulsingDot: {
    width: 9,
    height: 9,
    borderRadius: 100,
    position: 'absolute',
    opacity: 0.5,
  },
});
