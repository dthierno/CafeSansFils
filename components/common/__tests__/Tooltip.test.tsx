import React from "react";
import { fireEvent, render, screen } from "@testing-library/react-native";
import Tooltip from "@/components/common/Tooltip";
import { Globe } from "lucide-react-native";
import COLORS from "@/constants/Colors";


describe("Tooltip Component", () => {
  const mockOnPress = jest.fn();

  it("renders correctly with default props", () => {
    render(<Tooltip label="Default Tooltip" />);

    // Verify the label is rendered
    expect(screen.getByText("Default Tooltip")).toBeTruthy();

    // Verify the chevron is rendered by default
    expect(screen.getAllByTestId("chevron-down")).toBeTruthy();
  });

  it("renders with the correct label", () => {
    const label = "Custom Tooltip";
    render(<Tooltip label={label} />);

    // Verify the custom label is displayed
    expect(screen.getByText(label)).toBeTruthy();
  });

  it("renders the correct icon and color based on status", () => {
    render(<Tooltip label="Green Status" status="green" />);
    const greenIcon = screen.getAllByTestId("tooltip-icon");

    // Verify the icon is rendered with the correct color
    expect(greenIcon[0].props.fill).toBe(COLORS.status.green);

    render(<Tooltip label="Orange Status" status="orange" />);
    const orangeIcon = screen.getAllByTestId("tooltip-icon");
    expect(orangeIcon[0].props.fill).toBe(COLORS.status.orange);

    render(<Tooltip label="Red Status" status="red" />);
    const redIcon = screen.getAllByTestId("tooltip-icon");
    expect(redIcon[0].props.fill).toBe(COLORS.status.red);
  });

  it("does not render the chevron when showChevron is false", () => {
    render(<Tooltip label="No Chevron" showChevron={false} />);

    // Verify the chevron is not rendered
    expect(screen.queryByTestId("chevron-down")).toBeNull();
  });

  it("renders a custom icon", () => {
    render(<Tooltip label="Custom Icon" Icon={Globe} />);

    // Verify the custom icon is rendered
    const customIcon = screen.getAllByTestId("tooltip-icon");
    expect(customIcon).toBeTruthy();
  });

  it("triggers onPress callback when pressed", () => {
    render(<Tooltip label="Pressable Tooltip" onPress={mockOnPress} />);
    const tooltip = screen.getByTestId("tooltip-container");

    // Simulate a press
    fireEvent.press(tooltip);

    // Verify the onPress callback is triggered
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const tree = render(<Tooltip label="Snapshot Tooltip" />);
    expect(tree).toMatchSnapshot();
  });
});
