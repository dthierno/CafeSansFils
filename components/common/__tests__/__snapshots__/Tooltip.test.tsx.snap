// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Component matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F2F2F7",
      "borderRadius": 500,
      "flexDirection": "row",
      "gap": 4,
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 10,
    }
  }
  testID="tooltip-container"
>
  <Text
    style={
      [
        {
          "fontFamily": "Inter-SemiBold",
          "fontSize": 12,
        },
        {
          "color": "#000000",
        },
        false,
      ]
    }
  >
    Snapshot Tooltip
  </Text>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={16}
    bbWidth={16}
    fill="none"
    focusable={false}
    height={16}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="#000000"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2.8}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 16,
          "width": 16,
        },
      ]
    }
    testID="chevron-down"
    vbHeight={24}
    vbWidth={24}
    width={16}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2.8}
    >
      <RNSVGPath
        d="m6 9 6 6 6-6"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2.8}
        testID="chevron-down"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
